feat(ras-acc): support shipping, custom fields, and order notes in modal checkout

dkoo opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request, and the reason for such changes. -->

Adds support for a new option to show different billing fields for donations vs. non-donation transactions, and to show order notes in either case. Also adds support for shipping methods and fields as implemented in https://github.com/Automattic/newspack-blocks/pull/1626.

### How to test the changes in this Pull Request:

Check out this branch and https://github.com/Automattic/newspack-blocks/pull/1626.

#### Order notes

1. Visit **Newspack > Reader Revenue > Donations**. Scroll down to the "Donation Billing Fields" section and confirm there's slightly different copy and styles, as well as a `*` indicator next to fields that are required for checkout when shown.

<img width="1052" alt="Screenshot 2023-12-08 at 6 53 58 PM" src="https://github.com/Automattic/newspack-plugin/assets/2230142/399cffaf-75e7-4d64-9f6d-9450caa9d346">

2. Confirm there's a new checkbox option for "Order notes", disabled by default.
3. Check it and save.
4. As a reader, start a new donation and confirm you see the "Order notes" field on the first modal checkout screen:

<img width="771" alt="Screenshot 2023-12-08 at 6 41 34 PM" src="https://github.com/Automattic/newspack-plugin/assets/2230142/ae1ce8ab-74f2-47b2-8a1d-3e3a973cc656">

5. Fill out the Order Notes field and complete the transaction. In the admin, find the order and confirm that your note is attached to the order:

<img width="1252" alt="Screenshot 2023-12-08 at 6 42 29 PM" src="https://github.com/Automattic/newspack-plugin/assets/2230142/fcf333e5-c358-4651-a4a2-883ac5f57f6a">

#### Customized billing fields

1. Install and activate the the [Checkout Field Editor plugin](https://woo.com/products/woocommerce-checkout-field-editor/) (let me know if you need the package as it's a premium plugin).
2. In WP admin, visit **WooCommerce > Checkout Fields**. Make some customizations to the core billing fields, such as changing the labels and placeholders, as well as the "required" validation status for some fields (it will only let you make certain fields optional, such as Phone). Also add some of your own custom fields with various options. Save.

<img width="1334" alt="Screenshot 2023-12-08 at 6 51 35 PM" src="https://github.com/Automattic/newspack-plugin/assets/2230142/cd59bdec-0715-4a4d-9edb-21f7733a1d41">

3. Visit **Newspack > Reader Revenue > Donations**. Confirm that the available fields for both the Donation and Non-donation Billing Fields sections now reflect your customizations from step 2: core field labels will be as configured in the Checkout Fields plugin, and your custom fields will appear as additional fields that can be enabled or disabled like the other fields.
4. Enable the customized core fields and the new custom fields for both Donation and Non-donation Billing Fields and save.
5. As a reader, complete both a donation and a non-donation transaction and confirm that your customizations and custom fields are presented during checkout, and that you can complete both transactions as expected.

<img width="659" alt="Screenshot 2023-12-08 at 6 52 26 PM" src="https://github.com/Automattic/newspack-plugin/assets/2230142/3c9308bc-7599-4ab5-9d3c-57ad9ade0449">

### Other information:


* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205855958320407

dkoo committed:
feat: add support for order notes as a billing field

dkoo committed:
fix: remove unused import

leogermani reviewed (approved):
Everything works as expected. The custom fields validation also work. The only thing I noticed is that the validation message for the custom fields, created via the plugin:

* Show up at the top of the modal, not by the field as all others
* Are not translated, as all others

These should all be related to the plugin, so it's fine.

## Design review needed

I also want to flag that we need to rethink the Reader Revenue > Donations admin page. (And excuse me if this is already covered).

Not sure if in this PR, but certainly during this project, I think we should refactor how we present the options there a bit (before we work on the whole rearchitecture), because it's very confusing.

The page is about donations, but then all of a sudden there are "non-donation fields there".

Also, in the "Donation fields" we say "Configure which billing fields should be shown by the donation checkout form.". And then, just a sentence later: "Note that for shippable products, address fields will always be shown." It doesn't make a lot of sense.

We need to be explicit about where those billing fields apply -> The checkout button and the donation blocks.

Maybe have a "Billing fields" section of its own, separated from "Donations".

Is this flagged somewhere already?

miguelpeixe commented:
> The page is about donations, but then all of a sudden there are "non-donation fields there".

I agree. We should be more intentional about how we handle this field customization. "Non-donation fields" is too abstract of a feature and competes with products like the one mentioned (Checkout Field Editor).

As this is meant for the modal checkout, the place where we temporarily placed the "Donation fields" as the field selection for the checkout suggests a path forward:

https://github.com/Automattic/newspack-blocks/blob/115e9fb95c78a13f1d87f4d086e767311dc7007d/includes/class-modal-checkout.php#L603-L609

Having the field customization as a Checkout Button block feature is a more flexible approach. It adds complexity, as the validation must be stored separately, but makes more sense to me as a feature.

dkoo commented:
I hear you both, and I like the idea of approaching the field customization as a feature of the Checkout Button block. How about I remove that feature from this PR and we tackle that separately? I had mainly embarked on it here because it felt like an easy next step (since I was already in that part of the code for the Order Notes field).

Is the order notes field something that works well enough to keep here, in your opinion?

miguelpeixe commented:
Yep! I agree we can tackle it as a separate feature and the order notes field is a good addition to the existing "donate fields" custom selection.

dkoo commented:
@leogermani @miguelpeixe I've removed the feature to allow different billing fields for donations vs. non-donation transactions and updated testing instructions. Could use one more look, please!

leogermani commented:
Looking at this again now. There's a conflict there if you can have a look at it

leogermani commented:
@dkoo It seems you forgot to push your latest changes. The last commit is still from before my last comment

dkoo commented:
> @dkoo It seems you forgot to push your latest changes. The last commit is still from before my last comment

Whoops, you're right. I force-pushed to revert the non-donation billing commits.

leogermani reviewed (approved):
:+1:
